src/Audit.NET.PostgreSql/Providers/PostgreSqlDataProvider.cs

```diff
@@ -245,15 +244,89 @@ public IEnumerable<AuditEvent> EnumerateEvents(string whereExpression)
         /// <remarks>
         /// JSONB data querying: http://schinckel.net/2014/05/25/querying-json-in-postgres/, https://www.postgresql.org/docs/9.5/static/functions-json.html
         /// </remarks>
-        public IEnumerable<T> EnumerateEvents<T>(string whereExpression) where T : AuditEvent
+        private IEnumerable<T> EnumerateEvents<T>(string whereExpression) where T : AuditEvent
+        {
+            return this.EnumerateEvents<T>(whereExpression, string.Empty, string.Empty);
+        }
+        
+        /// <summary>
+        /// Provide paging functionality.
+        /// Note, the "Id"(ColumnName) will be the used in the Order-By.
+        /// </summary>
+        /// <param name="pageNumber">The specified page number.</param>
+        /// <param name="pageSize">The number of rows on a page</param>
+        /// <returns></returns>
+        public IEnumerable<AuditEvent> EnumerateEvents(int pageNumber, int pageSize)
+        {
+            return EnumerateEvents<AuditEvent>(pageNumber, pageSize);
+        }
+        
+        private IEnumerable<T> EnumerateEvents<T>(int pageNumber, int pageSize) where T : AuditEvent
+        {
+            return this.EnumerateEvents<T>(pageNumber, pageSize, string.Empty);
+        }
+        
+        /// <summary>
+        /// Provide paging functionality with a user-defined where-expression.
+        /// Note, the "Id"(ColumnName) will be the used in the Order-By.
+        /// </summary>
+        /// <param name="pageNumber">The specified page number.</param>
+        /// <param name="pageSize">The number of rows on a page</param>
+        /// <param name="whereExpression">where-clause that is applied before the "cut out the page" conditions are applied</param>
+        /// <returns></returns>
+        public IEnumerable<AuditEvent> EnumerateEvents(int pageNumber, int pageSize, string whereExpression)
+        {
+            return EnumerateEvents<AuditEvent>(pageNumber, pageSize, whereExpression);
+        }
+        
+        private IEnumerable<T> EnumerateEvents<T>(int pageNumber, int pageSize, string whereExpression) where T : AuditEvent
+        {
+            int safePageNumber = Math.Max(pageNumber, 1);
+            int safePageSize = Math.Max(pageSize, 1);
+            int offset = (safePageSize * safePageNumber) - safePageSize;
+            string schema = GetSchema(null);
+            
+            /* note, the "where-clause" must be applied "inside" the "derived1" inner-query...thus the variable name "inner-where-clause" */
+            string innerWhereClause = string.IsNullOrWhiteSpace(whereExpression) ? "" : $" WHERE {whereExpression}";
+
+            /* create the paging-query...which uses an inner-derived-table to capture the ROW_NUMBER values */
```

Comment:
here is the major change.  the ROW_NUMBER is a little more involved than a simple SELECT-FROM-WHERE query.

But I tested it with "gaps" in the values of the PrimaryKey(s) (aka 'Id-Column').